gapis/trace/tracer/tracer.go

```diff
@@ -49,6 +51,17 @@ type Process interface {
 	Capture(ctx context.Context, start task.Signal, stop task.Signal, w io.Writer, written *int64) (size int64, err error)
 }
 
+type GpuCounter struct {
+	Id    uint32
+	Name  string
+	Check func(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool
+}
+
+type Validator interface {
+	Validate(ctx context.Context, processor *perfetto.Processor) error
+	GetCounterIds() []uint32
+}
+
```

Comment:
there will be `arm` package using `GpuCounter` type so I put it at the same level of `Validator`

Reply:
Since `trace/android` depends on trace/android/adreno, `GpuCounter` can't be moved to `trace/android`. Not sure whether there's a different place I can move it, but if we keep it in `tracer`, then I would prefer to keep `Validator` in `tracer` as well.

Reply:
I think the packages should be:
- `trace/android/validate` only base deps (contains `Validator` interface and shared code)
- `trace/android/adreno` only base deps and `trace/android/validate`
- `trace/android/arm` only base deps and `trace/android/validate`
- `trace/android` only base deps, `trace/android/validate`, and the impls

If you want to share the counter objects, you can share the entire counter verification, no?

Reply:
thanks! Will do a dedicated `validate/` package.

hmmmmm, not sure what do you mean by counter objects. `mali` and `adreno` expose different counters so no, the counter verification can't be shared. Maybe some checkers can but we can adjust when we see counters from `mali`.

Reply:
Mark as resolved.

Reply:
What I meant was that you can change the `Validator` interface to return the counter objects instead of just the IDs. You will still create the counters along with their checker functions inside each impl package, but executing the counter queries and all that can be shared.